beacon/goclient/attest.go

```diff
@@ -27,22 +27,106 @@ func (gc *GoClient) AttesterDuties(ctx context.Context, epoch phase0.Epoch, vali
 	return resp.Data, nil
 }
 
-func (gc *GoClient) GetAttestationData(slot phase0.Slot, committeeIndex phase0.CommitteeIndex) (*phase0.AttestationData, spec.DataVersion, error) {
+// GetAttestationData returns attestation data for a given slot (which is same for all 64 committeeIndex
+// values that identify Ethereum committees chosen to attest on this slot).
+// Note, committeeIndex is an optional parameter that will be used to set AttestationData.Index
+// in the resulting data returned from this function.
+// Note, result returned is meant to be read-only, it's not safe to modify it (because it will be
+// accessed by multiple concurrent readers).
+func (gc *GoClient) GetAttestationData(slot phase0.Slot, committeeIndex phase0.CommitteeIndex) (
+	result *phase0.AttestationData,
+	version spec.DataVersion,
+	err error,
+) {
+	// Final processing for result returned.
+	defer func() {
+		if err != nil {
+			// Nothing to process, just propagate error.
+			return
+		}
+
+		// Assign committeeIndex passed to GetAttestationData call, the rest of attestation data stays
+		// unchanged.
+		// Note, we cannot return result object directly here modifying its Index value because it
+		// would be unsynchronised concurrent write (since it's cached for other concurrent readers
+		// to access). Hence, we return shallow copy here. We don't need to return deep copy because
+		// the callers of GetAttestationData will only use this data to read it (they won't update it).
+		result = &phase0.AttestationData{
+			Slot:            result.Slot,
+			Index:           committeeIndex,
+			BeaconBlockRoot: result.BeaconBlockRoot,
+			Source:          result.Source,
+			Target:          result.Target,
+		}
+	}()
+
+	// Check cache.
+	cachedResult, ok := gc.attestationDataCache.Get(slot)
+	if ok {
+		return cachedResult, spec.DataVersionPhase0, nil
+	}
+
+	// Have to make beacon node request and cache the result.
 	attDataReqStart := time.Now()
-	resp, err := gc.client.AttestationData(gc.ctx, &api.AttestationDataOpts{
-		Slot:           slot,
-		CommitteeIndex: committeeIndex,
-	})
+	result, err = func() (*phase0.AttestationData, error) {
```

Comment:
Can we extract this into another method?

Reply:
We could, but since `GetAttestationData` is quite short and simple (especially due to other improvements we've now applied) it would probably be harder to read if we do it,

I guess there is some balance to strive for when it comes to:
- having short functions
- having too many of them (having to jump around a lot when reading code)

so here I think it's a good example of "keeping it all in one place" being easier to read.

Reply:
okay, now this method is smaller, so I don't mind keeping this as is